e3nn_jax/_src/reduced_tensor_product.py

```diff
@@ -281,9 +205,189 @@ def _reduced_tensor_product_basis(
+        We need to pad this term to have shape (ir.dim, ..., ir.dim, ...) instead.
+        A similar logic applies for all of the terms.
+        """
+        inp_irreps_dims = [ir.dim for ir in irreps]
+        inp_irreps_dims_cumsum_before = cumsum_before(inp_irreps_dims)
+        inp_irreps_dims_cumsum_after = cumsum_after(inp_irreps_dims)
+
+        def compute_padding_for_irrep_index(irrep_index: int):
+            dims_before = inp_irreps_dims_cumsum_before[irrep_index]
+            dims_after = inp_irreps_dims_cumsum_after[irrep_index]
+            return (dims_before, dims_after)
+
+        return [compute_padding_for_irrep_index(irrep_index) for irrep_index in irrep_indices] + [(0, 0)]
+
+    def repeat_indices(indices: Sequence[int], powers: Sequence[int]) -> List[int]:
+        """Given [i1, i2, ...] and [p1, p2, ...], returns [i1, i1, ... (p1 times), i2, i2, ... (p2 times), ...]"""
+        repeated_indices = []
+        for index, power in zip(indices, powers):
+            repeated_indices.extend([index] * power)
+        return repeated_indices
+
+    def generate_permutations(seq: Sequence[float]) -> Iterator[Tuple[Sequence[float], Sequence[int]]]:
+        """Generates permutations of a sequence along with the indices used to create the permutation."""
+        indices = range(len(seq))
+        for permuted_indices in itertools.permutations(indices):
+            permuted_sequence = tuple(seq[index] for index in permuted_indices)
+            yield permuted_sequence, permuted_indices
+
+    def cumsum_before(seq: Sequence[float]) -> np.ndarray:
+        """Returns the cumulative sum before every index.
+
+        For example, cumsum_before([1, 2, 3]) == [0, 1, 3].
+        """
+        return np.cumsum([0, *seq])[:-1]
+
+    def cumsum_after(seq: Sequence[float]) -> np.ndarray:
+        """Returns the cumulative sum after every index.
+
+        For example, cumsum_after([1, 2, 3]) == [5, 3, 0].
+        """
+        return cumsum_before(seq[::-1])[::-1]
+
+    def reshape_for_basis_product(terms: Sequence[e3nn.IrrepsArray], non_zero_powers: Sequence[float]):
+        """Adds extra axes to each term to be compatible for reduce_basis_product()."""
+        term_powers_cumsum_before = cumsum_before(non_zero_powers)
+        term_powers_cumsum_after = cumsum_after(non_zero_powers)
+
+        def reshape_term_for_basis_product(index, term):
+            new_shape = (
+                (1,) * term_powers_cumsum_before[index]
+                + term.shape[:-1]
+                + (1,) * term_powers_cumsum_after[index]
+                + term.shape[-1:]
+            )
+            return term.reshape(new_shape)
+
+        return [reshape_term_for_basis_product(index, term) for index, term in enumerate(terms)]
+
+    irreps = e3nn.Irreps(irreps)
+
+    # Precompute powers of irreps.
+    irreps_powers = {}
+    for i, mul_ir in enumerate(irreps):
+        # TODO what if multiplicity is large here?
```

Comment:
@ameya98 can you fix this case?

If the mul > 1 we should not call the default algo but deal with the mul ourself in this function. Do you see what I mean?